asterius/src/Asterius/Types/SymbolSet.hs

```diff
@@ -0,0 +1,123 @@
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- |
+-- Module      :  Asterius.Types.SymbolSet
+-- Copyright   :  (c) 2018 EURL Tweag
+-- License     :  All rights reserved (see LICENCE file in the distribution).
+--
+-- The 'SymbolSet' type represents a set of elements of type 'EntitySymbol'.
+-- Internally it is represented as an 'IM.IntMap', mapping the keys of the
+-- uniques of the 'EntitySymbol's to the 'EntitySymbol's themselves.
+module Asterius.Types.SymbolSet
+  ( -- * SymbolSet type
+    SymbolSet,
+
+    -- * Construction
+    empty,
+    singleton,
+
+    -- * Query
+    null,
+
+    -- * Combine
+    union,
+    unions,
+    difference,
+
+    -- * Folds and Maps
+    foldr',
+
+    -- * Conversion
+    toIntSet,
+
+    -- ** Lists
+    fromList,
+    toList,
+  )
+where
+
+import Asterius.Types.EntitySymbol
+import Data.Data
+import qualified Data.Foldable as Foldable
+import qualified Data.IntMap.Lazy as IM
```

Comment:
We should be using `Data.IntMap.Strict` for `SymbolSet`, since the symbols are never lazy.

Reply:
Ah, right. Fixed.